packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/index.tsx

```diff
@@ -0,0 +1,238 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { useState } from 'react';
+import { useForm } from 'react-hook-form';
+import clsx from 'clsx';
+import { useTranslation } from 'react-i18next';
+import { Collapse } from '@mui/material';
+import styles from './styles.module.scss';
+import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowUp';
+import { ReactComponent as SwapIcon } from '../../../../assets/images/swap.svg';
+import TextButton from '../../../Form/Button/TextButton';
+import Switch from '../../../Form/Switch';
+import Unit from '../../../Form/Unit';
+import NumberInput from '../../../Form/NumberInput';
+import {
+  location_area,
+  soilAmounts,
+  soilAmountsVolume,
+  applicationRateWeight,
+  applicationRateVolume,
+} from '../../../../util/convert-units/unit';
+import { TASK_PRODUCT_FIELD_NAMES, type TaskProductFormFields } from '../types';
+import { AreaApplicationSummary } from './AreaApplicationSummary';
+import { useQuantityApplicationRate } from './useQuantityApplicationRate';
+
+const SOIL_AMENDMENT_UNITS = {
+  WEIGHT: {
+    units: soilAmounts, // original
+  },
+  VOLUME: {
+    units: soilAmountsVolume,
+  },
+  APPLICATION_RATE_WEIGHT: {
+    units: applicationRateWeight,
+  },
+  APPLICATION_RATE_VOLUME: {
+    units: applicationRateVolume,
+  },
+};
+
+export interface Location {
```

Comment:
For some reason this feels like it belongs in a new types file in `/AreaDetails`

Reply:
How about a new `types.ts` file that lives next to the slices (since this is a return from a selector and not specific to any one part of the app)? If so it would be a very partial type -- with a comment to that effect -- because the actual type of Location is a crazy union of location table + area/point + barn/garden/field etc and I don't even want to start to type that (and under Anto we agreed to not write types without immediate functional use).

Alternatively, I think it could also be kept here but renamed... `ApplicationRateLocation` maybe? To indicate it's only the properties necessary in `<QuantityApplicationRate />`?

What do you think?

Reply:
No action needed, but if you think a types file makes sense somewhere else that it might be reused then I am all for that.